File: ts/src/resource_report.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import {
  recommendAdaptiveResources,
  type AdaptiveResourceRecommendation,
  type ResourceObservation,
  type ResourcePolicyConfig,
  type ResourceSpec
} from "./resource_policy";
import type { ProviderName } from "./types";

type Args = {
  inputs: string[];
  resultsDir?: string;
  output?: string;
  suggestedConfigOutput?: string;
  format: "json" | "markdown";
  minSamples: number;
};

type ObservationGroup = {
  key: string;
  provider: ProviderName;
  env_type?: string;
  repo_key?: string;
  source_id?: string;
  runtime?: string;
  image_id?: string;
  image_version?: string;
  manifest_hash?: string;
  observations: ResourceObservation[];
};

type AggregatedResourceSuggestion = {
  key: string;
  provider: ProviderName;
  env_type?: string;
  repo_key?: string;
  source_id?: string;
  runtime?: string;
  image_id?: string;
  image_version?: string;
  manifest_hash?: string;
  sample_count: number;
  pass_count: number;
  failure_counts: Record<string, number>;
  p95_wall_seconds?: number;
  p95_peak_rss_gb?: number;
  p95_disk_gb?: number;
  p95_prepare_seconds?: number;
  p95_verify_seconds?: number;
  recommended: ResourceSpec;
  confidence: "low" | "medium" | "high";
  reasons: string[];
};

function parseArgs(argv: string[]): Args {
  const values = new Map<string, string>();
  for (let index = 0; index < argv.length; index += 2) {
    values.set(argv[index], argv[index + 1]);
  }
  const inputText = values.get("--input") ?? "";
  return {
    inputs: inputText
      .split(",")
      .map((item) => item.trim())
      .filter(Boolean),
    resultsDir: values.get("--results-dir"),
    output: values.get("--output"),
    suggestedConfigOutput: values.get("--suggested-config-output"),
    format: parseFormat(values.get("--format") ?? "markdown"),
    minSamples: Number.parseInt(values.get("--min-samples") ?? "2", 10)
  };
}

function parseFormat(value: string): Args["format"] {
  if (value === "json" || value === "markdown") {
    return value;
  }
  throw new Error(`Unsupported resource report format: ${value}`);
}

function loadObservations(args: Args): ResourceObservation[] {
  const paths = [...args.inputs];
  if (args.resultsDir) {
    paths.push(
      ...readdirSync(args.resultsDir)
        .filter((file) => file.endsWith(".json") || file.endsWith(".jsonl"))
        .map((file) => join(args.resultsDir as string, file))
    );
  }
  const observations = paths.flatMap((path) => loadObservationFile(resolve(path)));
  return observations.filter(isResourceObservation);
}

function loadObservationFile(path: string): unknown[] {
  if (!existsSync(path)) {
    throw new Error(`Resource observation input not found: ${path}`);
  }
  const text = readFileSync(path, "utf8").trim();
  if (!text) {
    return [];
  }
  if (path.endsWith(".jsonl")) {
    return text
      .split("\n")
      .map((line) => line.trim())
      .filter(Boolean)
      .map((line) => JSON.parse(line));
  }
  const parsed = JSON.parse(text) as unknown;
  if (Array.isArray(parsed)) {
    return parsed;
  }
  if (isObject(parsed)) {
    const direct = parsed.resource_observation;
    const results = Array.isArray(parsed.results) ? parsed.results : [];
    return [
      ...(direct ? [direct] : []),
      ...results.map((item) => (isObject(item) ? item.resource_observation : undefined)).filter(Boolean)
    ];
  }
  return [];
}

function aggregateObservations(observations: ResourceObservation[]): AggregatedResourceSuggestion[] {
  return [...groupObservations(observations).values()]
    .map(aggregateGroup)
    .sort((left, right) => left.provider.localeCompare(right.provider) || left.key.localeCompare(right.key));
}

function groupObservations(observations: ResourceObservation[]): Map<string, ObservationGroup> {
  const groups = new Map<string, ObservationGroup>();
  for (const observation of observations) {
    const key = groupKey(observation);
    const existing = groups.get(key);
    if (existing) {
      existing.observations.push(observation);
      continue;
    }
    groups.set(key, {
      key,
      provider: observation.provider,
      env_type: observation.env_type,
      repo_key: observation.repo_key,
      source_id: observation.source_id,
      runtime: observation.runtime,
      image_id: observation.image_id,
      image_version: observation.image_version,
      manifest_hash: observation.manifest_hash,
      observations: [observation]
    });
  }
  return groups;
}

function groupKey(observation: ResourceObservation): string {
  return [
    observation.provider,
    observation.env_type ?? "unknown-env",
    observation.repo_key ?? observation.task_id ?? "unknown-scope",
    observation.runtime ?? "unknown-runtime",
    observation.image_id ?? observation.docker_image ?? "no-image",
    observation.manifest_hash ?? "no-manifest"
  ].join("|");
}

function aggregateGroup(group: ObservationGroup): AggregatedResourceSuggestion {
  const recommendations = group.observations.map((observation) => recommendAdaptiveResources(observation));
  const recommended = combineRecommendations(group.observations, recommendations);
  const failures = failureCounts(group.observations);
  const sampleCount = group.observations.length;
  return {
    key: group.key,
    provider: group.provider,
    ...(group.env_type ? { env_type: group.env_type } : {}),
    ...(group.repo_key ? { repo_key: group.repo_key } : {}),
    ...(group.source_id ? { source_id: group.source_id } : {}),
    ...(group.runtime ? { runtime: group.runtime } : {}),
    ...(group.image_id ? { image_id: group.image_id } : {}),
    ...(group.image_version ? { image_version: group.image_version } : {}),
    ...(group.manifest_hash ? { manifest_hash: group.manifest_hash } : {}),
    sample_count: sampleCount,
    pass_count: group.observations.filter((observation) => observation.passed).length,
    failure_counts: failures,
    p95_wall_seconds: percentile(group.observations.map((observation) => observation.usage.wall_seconds).filter(isFiniteNumber), 95),
    p95_peak_rss_gb: percentile(group.observations.map((observation) => observation.usage.peak_rss_gb).filter(isFiniteNumber), 95),
    p95_disk_gb: percentile(group.observations.map((observation) => observation.disk_usage?.total_gb).filter(isFiniteNumber), 95),
    p95_prepare_seconds: percentile(group.observations.map((observation) => observation.phase_seconds?.prepare_seconds).filter(isFiniteNumber), 95),
    p95_verify_seconds: percentile(group.observations.map((observation) => observation.phase_seconds?.verify_seconds).filter(isFiniteNumber), 95),
    recommended,
    confidence: confidenceFor(sampleCount, failures),
    reasons: [...new Set(recommendations.flatMap((recommendation) => recommendation.reasons))]
  };
}

function combineRecommendations(observations: ResourceObservation[], recommendations: AdaptiveResourceRecommendation[]): ResourceSpec {
  const recommended = recommendations.map((recommendation) => recommendation.recommended);
  const diskFromUsage = observations
    .map((observation) => observation.disk_usage?.total_gb)
    .filter(isFiniteNumber)
    .map((gb) => roundUpTier(gb * 1.4, [10, 20, 40, 80]));
  return {
    cpu: Math.max(...recommended.map((item) => item.cpu)),
    memoryGb: Math.max(...recommended.map((item) => item.memoryGb)),
    diskGb: Math.max(...recommended.map((item) => item.diskGb), ...diskFromUsage),
    timeoutSeconds: Math.max(...recommended.map((item) => item.timeoutSeconds))
  };
}

function failureCounts(observations: ResourceObservation[]): Record<string, number> {
  const counts: Record<string, number> = {};
  for (const observation of observations) {
    counts[observation.failure_class] = (counts[observation.failure_class] ?? 0) + 1;
  }
  return counts;
}

function confidenceFor(sampleCount: number, failures: Record<string, number>): AggregatedResourceSuggestion["confidence"] {
  const resourceFailures = (failures.memory_limit ?? 0) + (failures.disk_full ?? 0) + (failures.cpu_limit ?? 0);
  if (sampleCount >= 5 && resourceFailures === 0) {
    return "high";
  }
  if (sampleCount >= 2) {
    return "medium";
  }
  return "low";
}

function suggestedConfig(rows: AggregatedResourceSuggestion[], minSamples: number): ResourcePolicyConfig {
  const config: ResourcePolicyConfig = {
    schema_version: 1,
    default_policy: "adaptive",
    provider_defaults: {},
    env_type_defaults: {},
    repo_overrides: {}
  };
  for (const row of rows.filter((item) => item.sample_count >= minSamples)) {
    if (row.repo_key) {
      config.repo_overrides ??= {};
      config.repo_overrides[row.repo_key] ??= {};
      config.repo_overrides[row.repo_key][row.provider] = row.recommended;
      continue;
    }
    if (row.env_type) {
      config.env_type_defaults ??= {};
      config.env_type_defaults[row.env_type] ??= {};
      config.env_type_defaults[row.env_type][row.provider] = row.recommended;
    }
  }
  return config;
}

function markdownReport(rows: AggregatedResourceSuggestion[], observationCount: number, minSamples: number): string {
  return [
    "# Resource Observation Report",
    "",
    `Generated: ${new Date().toISOString()}`,
    "",
    `Observations: ${observationCount}`,
    `Suggested config min samples: ${minSamples}`,
    "",
    "provider | scope | samples | passed | failures | p95 wall | p95 rss GB | p95 disk GB | recommended",
    "--- | --- | ---: | ---: | --- | ---: | ---: | ---: | ---",
    ...rows.map((row) =>
      [
        row.provider,
        row.repo_key ?? row.env_type ?? "unknown",
        String(row.sample_count),
        String(row.pass_count),
        failureSummary(row.failure_counts),
        formatNumber(row.p95_wall_seconds),
        formatNumber(row.p95_peak_rss_gb),
        formatNumber(row.p95_disk_gb),
        formatSpec(row.recommended)
      ].join(" | ")
    ),
    ""
  ].join("\n");
}

function failureSummary(counts: Record<string, number>): string {
  return Object.entries(counts)
    .filter(([, count]) => count > 0)
    .map(([failure, count]) => `${failure}:${count}`)
    .join(", ");
}

function formatSpec(spec: ResourceSpec): string {
  return `${spec.cpu} CPU / ${spec.memoryGb} GB / ${spec.diskGb} GB disk / ${spec.timeoutSeconds}s`;
}

function formatNumber(value: number | undefined): string {
  return value === undefined || Number.isNaN(value) ? "-" : value.toFixed(2);
}

function writeOutput(path: string | undefined, content: string): void {
  if (!path) {
    console.log(content);
    return;
  }
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, content);
  console.log(`wrote ${path}`);
}

function percentile(values: number[], p: number): number | undefined {
  if (values.length === 0) {
    return undefined;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.ceil((p / 100) * sorted.length) - 1);
  return sorted[index];
}

function roundUpTier(value: number, tiers: number[]): number {
  return tiers.find((tier) => tier >= value) ?? tiers.at(-1) ?? Math.ceil(value);
}

function isResourceObservation(value: unknown): value is ResourceObservation {
  return (
    isObject(value) &&
    typeof value.provider === "string" &&
    typeof value.resource_policy === "string" &&
    isObject(value.requested) &&
    isObject(value.effective) &&
    isObject(value.usage) &&
    typeof value.return_code === "number" &&
    typeof value.passed === "boolean" &&
    typeof value.failure_class === "string"
  );
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isFiniteNumber(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

async function main(): Promise<void> {
  const args = parseArgs(Bun.argv.slice(2));
  const observations = loadObservations(args);
  const rows = aggregateObservations(observations);
  const config = suggestedConfig(rows, args.minSamples);
  const payload = {
    schema_version: 1,
    generated_at: new Date().toISOString(),
    observation_count: observations.length,
    group_count: rows.length,
    min_samples: args.minSamples,
    groups: rows,
    suggested_config: config
  };
  const content = args.format === "json" ? `${JSON.stringify(payload, null, 2)}\n` : markdownReport(rows, observations.length, args.minSamples);
  writeOutput(args.output, content);
  if (args.suggestedConfigOutput) {
    writeOutput(args.suggestedConfigOutput, `${JSON.stringify(config, null, 2)}\n`);
  }
}

if (import.meta.main) {
  await main();
}

File: py/code_sandbox_bench/resource_report.py
import argparse
import json
import math
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from code_sandbox_bench.resource_policy import recommend_adaptive_resources


def iso_now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", default="")
    parser.add_argument("--results-dir", type=Path)
    parser.add_argument("--output", type=Path)
    parser.add_argument("--suggested-config-output", type=Path)
    parser.add_argument("--format", choices=["json", "markdown"], default="markdown")
    parser.add_argument("--min-samples", type=int, default=2)
    return parser.parse_args()


def load_observations(args: argparse.Namespace) -> list[dict[str, Any]]:
    paths = [Path(item) for item in args.input.split(",") if item.strip()]
    if args.results_dir:
        paths.extend(path for path in args.results_dir.iterdir() if path.suffix in {".json", ".jsonl"})
    observations: list[dict[str, Any]] = []
    for path in paths:
        observations.extend(load_observation_file(path))
    return [item for item in observations if is_resource_observation(item)]


def load_observation_file(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        raise FileNotFoundError(f"Resource observation input not found: {path}")
    text = path.read_text(encoding="utf-8").strip()
    if not text:
        return []
    if path.suffix == ".jsonl":
        return [json.loads(line) for line in text.splitlines() if line.strip()]
    parsed = json.loads(text)
    if isinstance(parsed, list):
        return [item for item in parsed if isinstance(item, dict)]
    if isinstance(parsed, dict):
        observations = []
        if isinstance(parsed.get("resource_observation"), dict):
            observations.append(parsed["resource_observation"])
        if isinstance(parsed.get("results"), list):
            observations.extend(
                item["resource_observation"]
                for item in parsed["results"]
                if isinstance(item, dict) and isinstance(item.get("resource_observation"), dict)
            )
        return observations
    return []


def aggregate_observations(observations: list[dict[str, Any]]) -> list[dict[str, Any]]:
    groups: dict[str, dict[str, Any]] = {}
    for observation in observations:
        key = group_key(observation)
        group = groups.setdefault(
            key,
            {
                "key": key,
                "provider": observation.get("provider"),
                "env_type": observation.get("env_type"),
                "repo_key": observation.get("repo_key"),
                "source_id": observation.get("source_id"),
                "runtime": observation.get("runtime"),
                "image_id": observation.get("image_id"),
                "image_version": observation.get("image_version"),
                "manifest_hash": observation.get("manifest_hash"),
                "observations": [],
            },
        )
        group["observations"].append(observation)
    return sorted((aggregate_group(group) for group in groups.values()), key=lambda row: (str(row.get("provider")), row["key"]))


def group_key(observation: dict[str, Any]) -> str:
    return "|".join(
        [
            str(observation.get("provider") or "unknown-provider"),
            str(observation.get("env_type") or "unknown-env"),
            str(observation.get("repo_key") or observation.get("task_id") or "unknown-scope"),
            str(observation.get("runtime") or "unknown-runtime"),
            str(observation.get("image_id") or observation.get("docker_image") or "no-image"),
            str(observation.get("manifest_hash") or "no-manifest"),
        ]
    )


def aggregate_group(group: dict[str, Any]) -> dict[str, Any]:
    observations = group["observations"]
    recommendations = [recommend_adaptive_resources(observation) for observation in observations]
    failures = failure_counts(observations)
    row = {
        "key": group["key"],
        "provider": group.get("provider"),
        "sample_count": len(observations),
        "pass_count": sum(1 for observation in observations if observation.get("passed") is True),
        "failure_counts": failures,
        "p95_wall_seconds": percentile([number_at(observation, "usage", "wall_seconds") for observation in observations], 95),
        "p95_peak_rss_gb": percentile([number_at(observation, "usage", "peak_rss_gb") for observation in observations], 95),
        "p95_disk_gb": percentile([number_at(observation, "disk_usage", "total_gb") for observation in observations], 95),
        "p95_prepare_seconds": percentile([number_at(observation, "phase_seconds", "prepare_seconds") for observation in observations], 95),
        "p95_verify_seconds": percentile([number_at(observation, "phase_seconds", "verify_seconds") for observation in observations], 95),
        "recommended": combine_recommendations(observations, recommendations),
        "confidence": confidence_for(len(observations), failures),
        "reasons": sorted({reason for recommendation in recommendations for reason in recommendation.get("reasons", [])}),
    }
    for key in ("env_type", "repo_key", "source_id", "runtime", "image_id", "image_version", "manifest_hash"):
        if group.get(key):
            row[key] = group[key]
    return row


def combine_recommendations(observations: list[dict[str, Any]], recommendations: list[dict[str, Any]]) -> dict[str, int]:
    recommended = [item["recommended"] for item in recommendations if isinstance(item.get("recommended"), dict)]
    disk_from_usage = [
        round_up_tier(value * 1.4, [10, 20, 40, 80])
        for value in [number_at(observation, "disk_usage", "total_gb") for observation in observations]
        if is_finite_number(value)
    ]
    return {
        "cpu": max(int(item.get("cpu") or 1) for item in recommended),
        "memoryGb": max(int(item.get("memoryGb") or 1) for item in recommended),
        "diskGb": max([int(item.get("diskGb") or 10) for item in recommended] + disk_from_usage),
        "timeoutSeconds": max(int(item.get("timeoutSeconds") or 180) for item in recommended),
    }


def failure_counts(observations: list[dict[str, Any]]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for observation in observations:
        key = str(observation.get("failure_class") or "none")
        counts[key] = counts.get(key, 0) + 1
    return counts


def confidence_for(sample_count: int, failures: dict[str, int]) -> str:
    resource_failures = failures.get("memory_limit", 0) + failures.get("disk_full", 0) + failures.get("cpu_limit", 0)
    if sample_count >= 5 and resource_failures == 0:
        return "high"
    if sample_count >= 2:
        return "medium"
    return "low"


def suggested_config(rows: list[dict[str, Any]], min_samples: int) -> dict[str, Any]:
    config: dict[str, Any] = {"schema_version": 1, "default_policy": "adaptive", "provider_defaults": {}, "env_type_defaults": {}, "repo_overrides": {}}
    for row in rows:
        if int(row["sample_count"]) < min_samples:
            continue
        provider = str(row["provider"])
        if row.get("repo_key"):
            config["repo_overrides"].setdefault(row["repo_key"], {})[provider] = row["recommended"]
        elif row.get("env_type"):
            config["env_type_defaults"].setdefault(row["env_type"], {})[provider] = row["recommended"]
    return config


def markdown_report(rows: list[dict[str, Any]], observation_count: int, min_samples: int) -> str:
    lines = [
        "# Resource Observation Report",
        "",
        f"Generated: {iso_now()}",
        "",
        f"Observations: {observation_count}",
        f"Suggested config min samples: {min_samples}",
        "",
        "provider | scope | samples | passed | failures | p95 wall | p95 rss GB | p95 disk GB | recommended",
        "--- | --- | ---: | ---: | --- | ---: | ---: | ---: | ---",
    ]
    for row in rows:
        lines.append(
            " | ".join(
                [
                    str(row.get("provider") or "-"),
                    str(row.get("repo_key") or row.get("env_type") or "unknown"),
                    str(row["sample_count"]),
                    str(row["pass_count"]),
                    failure_summary(row["failure_counts"]),
                    format_number(row.get("p95_wall_seconds")),
                    format_number(row.get("p95_peak_rss_gb")),
                    format_number(row.get("p95_disk_gb")),
                    format_spec(row["recommended"]),
                ]
            )
        )
    return "\n".join(lines + [""])


def is_resource_observation(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and isinstance(value.get("provider"), str)
        and isinstance(value.get("resource_policy"), str)
        and isinstance(value.get("requested"), dict)
        and isinstance(value.get("effective"), dict)
        and isinstance(value.get("usage"), dict)
        and isinstance(value.get("return_code"), int)
        and isinstance(value.get("passed"), bool)
        and isinstance(value.get("failure_class"), str)
    )


def number_at(value: dict[str, Any], key: str, nested_key: str) -> float | None:
    nested = value.get(key)
    item = nested.get(nested_key) if isinstance(nested, dict) else None
    return float(item) if is_finite_number(item) else None


def percentile(values: list[float | None], p: int) -> float | None:
    numbers = sorted(float(value) for value in values if is_finite_number(value))
    if not numbers:
        return None
    index = min(len(numbers) - 1, math.ceil((p / 100) * len(numbers)) - 1)
    return numbers[index]


def round_up_tier(value: float, tiers: list[int]) -> int:
    return next((tier for tier in tiers if tier >= value), tiers[-1] if tiers else math.ceil(value))


def is_finite_number(value: Any) -> bool:
    return isinstance(value, (int, float)) and math.isfinite(value)


def failure_summary(counts: dict[str, int]) -> str:
    return ", ".join(f"{key}:{count}" for key, count in counts.items() if count > 0)


def format_spec(spec: dict[str, Any]) -> str:
    return f"{spec['cpu']} CPU / {spec['memoryGb']} GB / {spec['diskGb']} GB disk / {spec['timeoutSeconds']}s"


def format_number(value: Any) -> str:
    return "-" if not is_finite_number(value) else f"{float(value):.2f}"


def write_output(path: Path | None, content: str) -> None:
    if path is None:
        print(content)
        return
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")
    print(f"wrote {path}")


def main() -> None:
    args = parse_args()
    observations = load_observations(args)
    rows = aggregate_observations(observations)
    config = suggested_config(rows, args.min_samples)
    payload = {
        "schema_version": 1,
        "generated_at": iso_now(),
        "observation_count": len(observations),
        "group_count": len(rows),
        "min_samples": args.min_samples,
        "groups": rows,
        "suggested_config": config,
    }
    content = json.dumps(payload, indent=2) + "\n" if args.format == "json" else markdown_report(rows, len(observations), args.min_samples)
    write_output(args.output, content)
    if args.suggested_config_output:
        write_output(args.suggested_config_output, json.dumps(config, indent=2) + "\n")


if __name__ == "__main__":
    main()
